### ISSUE TITLE
The es-proxy-deploy pods periodically stop responding and do not seem to recover

### ISSUE THREAD

**Issue by jwiss** [IBM/event-streams#121]
## Issue Description

* Describe the problem

The es-proxy-deploy pods periodically stop responding and do not seem to recover

* Describe the application flow or what steps were being taken when the problem occurred

The deployed system is for functional testing. There is no special setup. The product is deployed and runs fine until the error occurs.

* If the issue is reproducible provide a set of instructions describing how to reproduce it, including screenshots if they might assist

The issue occurs after the system has been deployed for a period of time. It looks like it is related to recovery from an unexpected network error. Details follow.

* Include any error messages, stack traces and logs associated with the error, in full, in text format (ie not a screenshot)

**Symptom**
On startup, The es-proxy-deploy pods monitor the ConfigMap es-staging-ibm-es-proxy-cm using a Kubernetes k8s.Client. See detail (1) below. The k8s.Client watches the ConfigMap and periodically receives an unexpected, likely network, error. See detail (2). On perhaps some class of errors the watcher then does not recover. We see a retry attempt fail with a AuthError. See (3) below. The pods remain in a running state but the container never starts and hence no longer responds to client requests.

**Details** 
(1) ConfigMap monitoring initializes k8sclient.go:166] Attempting to watch for changes to config map (retry set for 5 seconds) : configmap name : es-staging-ibm-es-proxy-cm
(2) The k8s.Client watching the es-staging-ibm-es-proxy-cm ConfigMap receives an error.
{"ibm_datetime":"2020-10-24T03:01:48Z","logLevel":"ERROR","module":"KafkaProxy", "ibm_messageId":"IES","message":"IES: Kube watcher status change : http2: server sent GOAWAY and closed the connection; LastStreamID=187, ErrCode=NO_ERROR, debug=\"\""}
(3) The retry to watch receives an auth error
{"ibm_datetime":"2020-10-24T03:02:14Z","logLevel":"ERROR","module":"KafkaProxy", "ibm_messageId":"IES","message":"IES: Error creating watcher : kubernetes api: Failure 401 Unauthorized"}

* Run the log collection script available at https://github.com/IBM/charts/tree/master/stable/ibm-eventstreams-dev/ibm_cloud_pak/pak_extensions/support/get-logs.sh and attach the output to this issue

For those with access, logs have been loaded and the issue discussed here:  https://ibm-security.slack.com/archives/CB777ACRM/p1603733272446300.   We can repost here if necessary.

* Describe the expected behaviour

The es-proxy-deploy pods should recover from unexpected kube issues, again likely network related.

* Did this used to work? If so what has changed, for example was any maintenance of the system performed?

## Environment

* IBM Event Streams Version: 2019.4.2
* IBM Cloud Private (ICP) Version: 3.2.4, running Openshift 4.3.35
* Operating system of ICP install: Red Hat Enterprise Linux release 8.2
* Browser (for UI issues):

**Comment by nictownsend** [comment]
```
kubectl version
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.2+7279a4a", GitCommit:"7279a4a", GitTreeState:"clean", BuildDate:"2020-08-31T17:24:39Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"

oc version
Client Version: 4.3.35
Server Version: 4.3.35
Kubernetes Version: v1.16.2+7279a4a
```

**Comment by jwiss** [comment]
We've also seen the issue on
```
oc version  
Client Version: 4.4.6
Server Version: 4.3.13
Kubernetes Version: v1.16.2
kubectl version
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.2", GitCommit:"a02f27a", GitTreeState:"clean", BuildDate:"2020-04-13T12:04:13Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"
```
FWIW, in this case we. see the GOWAY as in (2) above, followed by in (3) not by an auth error but instead but a refused connection.

**Comment by EmmaHumber** [comment]
Internal issue 6422 raised to track resolution.

**Comment by EmmaHumber** [comment]
### Issue Description

As above

### Issue Resolution

On a network disconnect the es-proxy-deploy proxy ( the "kafka proxy", or "edge proxy") only attempted to reconnect only once.  If this reconnect failed, processing stopped.

The kafka proxy has been updated to retry the reconnect.

### Workaround

Delete the affected es-proxy-deploy pod.  This will restart the kafka proxy and remake the network connections.

### Fix details

Internal issue: 6422
Fix target : 2019.4.4

**Comment by jwiss** [comment]
A bit of follow up. We suspect that a primary reason for the issue is a network issue between a master node hosting the apiserver and a worker hosting eventstreams. We see this snippet in apiserver logs.
```
kube-apiserver-14 I1113 17:01:48.751387       1 node_authorizer.go:195] NODE DENY: unknown node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get secret es-staging/es-staging-ibm-es-access-controller-sa-token-wg5cq
kube-apiserver-master1.sys-test-04.cp.fyre.ibm.com kube-apiserver-14 I1113 17:01:48.752229       1 node_authorizer.go:195] NODE DENY: unknown node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get secret es-staging/sa-es-staging
kube-apiserver-master1.sys-test-04.cp.fyre.ibm.com kube-apiserver-14 I1113 17:01:48.757223       1 node_authorizer.go:195] NODE DENY: unknown node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get secret es-staging/insights-pull-secret
 I1113 17:01:48.757308       1 node_authorizer.go:195] NODE DENY: unknown node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get secret es-staging/es-staging-ibm-es-proxy-sa-token-wqz7n
kube-apiserver-14 I1113 17:01:48.757608       1 node_authorizer.go:195] NODE DENY: unknown node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get configmap es-staging/es-staging-ibm-es-metrics-cm
kube-apiserver-master1.sys-test-04.cp.fyre.ibm.com kube-apiserver-14 I1113 17:01:48.804691       1 node_authorizer.go:195] NODE DENY: node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get unknown secret es-staging/es-staging-ibm-es-proxy-sa-dockercfg-cxxh8
kube-apiserver-master1.sys-test-04.cp.fyre.ibm.com kube-apiserver-14 I1113 17:01:48.809901       1 node_authorizer.go:195] NODE DENY: node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get unknown configmap es-staging/es-staging-ibm-es-release-cm
kube-apiserver-master0.sys-test-04.cp.fyre.ibm.com kube-apiserver-master1.sys-test-04.cp.fyre.ibm.com kube-apiserver-14 I1113 17:01:48.810770       1 node_authorizer.go:195] NODE DENY: node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get unknown configmap es-staging/kafka-setup
kube-apiserver-master0.sys-test-04.cp.fyre.ibm.com kube-apiserver-master2.sys-test-04.cp.fyre.ibm.com kube-apiserver-14I1113 17:01:48.811074       1 node_authorizer.go:195] NODE DENY: node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get unknown secret es-staging/es-staging-ibm-es-ui-sa-token-dthdx
kube-apiserver-master1.sys-test-04.cp.fyre.ibm.com kube-apiserver-14 I1113 17:01:48.845529       1 node_authorizer.go:195] NODE DENY: node "worker3.sys-test-04.cp.fyre.ibm.com" cannot get unknown secret es-staging/es-staging-ibm-es-rest-proxy-sa-dockercfg-xkwr5 
```
The resolution here would be on the apiserver end. The es code can loop asking for a connection to access a configmap, but until the apiserver issue is rectified, it seems es will be stuck.
